[TZone] Add API to allow processes to set their own bucket count policy

msaboff opened the pull request:
#### 101108025a14ad07924c2fff0bdea725abc0103d
<pre>
[TZone] Add API to allow processes to set their own bucket count policy
<a href="https://bugs.webkit.org/show_bug.cgi?id=282697">https://bugs.webkit.org/show_bug.cgi?id=282697</a>
<a href="https://rdar.apple.com/problem/139367739">rdar://problem/139367739</a>

Reviewed by Yijia Huang.

Added a new TZoneHeapManager function, initWithBucketSettings(), for setting bucket counts different than the default.
After changing the bucket parameters, initWithBucketSettings will then initialize the TZone manager.
Note that initWithBucketSettings must be called before the first TZone allocation, typically early in the main()
function of a process.

Also changed the default bucket counts to 5 buckets for small types, 3 buckets for large types, but left the
maximum size of small types at 128 bytes.  The changing of these defaults is based on performance and memory use
testing.

* Source/bmalloc/bmalloc/TZoneHeapManager.cpp:
(bmalloc::api::TZoneHeapManager::setBucketParams):
(bmalloc::api::TZoneHeapManager::init):
(bmalloc::api::TZoneHeapManager::ensureSingleton):
* Source/bmalloc/bmalloc/TZoneHeapManager.h:
(bmalloc::api::TZoneHeapManager::singleton):
(bmalloc::api::TZoneHeapManager::initWithBucketSettings):

Canonical link: <a href="https://commits.webkit.org/286433@main">https://commits.webkit.org/286433@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0ddab340409f5e1e9ca32ca08f4c2247e66455e6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75369 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28219 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79842 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26633 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2600 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59155 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17371 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78436 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64755 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39517 "Passed tests") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13 "Build is being retried. Recent messages:OS: Sequoia (15.1), Xcode: 16.1; Running apply-patch; Checked out pull request; Running run-layout-tests-in-stress-mode; layout-tests (retry)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22251 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24960 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/68519 "Built successfully and passed tests") | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67750 "") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22589 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81325 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/74631 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2708 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1703 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67400 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2859 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64743 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66680 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/16720 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10634 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8794 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/96899 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11752 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2665 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21165 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2690 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3625 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2697 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/bmalloc/bmalloc/TZoneHeapManager.h:
```diff
@@ -175,6 +193,9 @@ class TZoneHeapManager {
     TZoneHeapRandomizeKey m_tzoneKey;
+    unsigned bucketsForSmallSizes { defaultBucketsForSmallSizes };
+    unsigned bucketsForLargeSizes { defaultBucketsForLargeSizes };
+    unsigned maxSmallSize = { defaultMaxSmallSize };
```
We can just do `unsigned maxSmallSize { defaultMaxSmallSize };`.

msaboff replied:
Yep.

hyjorc1 commented on Source/bmalloc/bmalloc/TZoneHeapManager.h:
```diff
@@ -147,18 +150,33 @@ class TZoneHeapManager {
 private:
     BEXPORT static void ensureSingleton();
 
+    BEXPORT void setBucketParams(unsigned smallSizeCount, unsigned largeSizeCount = 0, unsigned smallSizeLimit = 0);
```
Do we really need to define the default values for the arguments since the default cases never happens?

msaboff replied:
This is here if the caller only wants to set the small and large counts and not the max small size.  This will probably be common.

hyjorc1 commented on Source/bmalloc/bmalloc/TZoneHeapManager.h:
```diff
@@ -147,18 +150,33 @@ class TZoneHeapManager {
         return *theTZoneHeapManager;
     }
 
+    BINLINE static void initWithBucketSettings(unsigned smallSizeCount, unsigned largeSizeCount = 0, unsigned smallSizeLimit = 0)
```
This function is not used, right?

msaboff replied:
It isn't used.  I'll use it in a subsequent patch.

hyjorc1 reviewed (approved):
r=me

msaboff committed:
[TZone] Add API to allow processes to set their own bucket count policy
https://bugs.webkit.org/show_bug.cgi?id=282697
rdar://problem/139367739

Reviewed by Yijia Huang.

Added a new TZoneHeapManager function, initWithBucketSettings(), for setting bucket counts different than the default.
After changing the bucket parameters, initWithBucketSettings will then initialize the TZone manager.
Note that initWithBucketSettings must be called before the first TZone allocation, typically early in the main()
function of a process.

Also changed the default bucket counts to 5 buckets for small types, 3 buckets for large types, but left the
maximum size of small types at 128 bytes.  The changing of these defaults is based on performance and memory use
testing.

* Source/bmalloc/bmalloc/TZoneHeapManager.cpp:
(bmalloc::api::TZoneHeapManager::setBucketParams):
(bmalloc::api::TZoneHeapManager::init):
(bmalloc::api::TZoneHeapManager::ensureSingleton):
* Source/bmalloc/bmalloc/TZoneHeapManager.h:
(bmalloc::api::TZoneHeapManager::singleton):
(bmalloc::api::TZoneHeapManager::initWithBucketSettings):

Canonical link: https://commits.webkit.org/286433@main
